tooling/greybox_fuzzer/src/mutation/int.rs

```diff
@@ -0,0 +1,558 @@
+        BinaryIntOperationMutation::Xor => lhs_int ^ rhs_int,
+        BinaryIntOperationMutation::And => lhs_int & rhs_int,
+        BinaryIntOperationMutation::Or => lhs_int | rhs_int,
+    };
+    i128_to_field(result_int.as_(), width)
+}
+
+/// Perform an unsigned integer binary operation on 2 field elements according to `mutation_operation`. Get field as a result
+fn add_sub_xor_and_or_unsigned<
+    T: WrappingAdd
+        + WrappingSub
+        + HasBits
+        + BitXor<Output = T>
+        + BitAnd<Output = T>
+        + BitOr<Output = T>
+        + AsPrimitive<i128>
+        + PrimInt,
+>(
+    lhs: &FieldElement,
+    rhs: &FieldElement,
+    mutation_operation: BinaryIntOperationMutation,
+) -> FieldElement {
+    let width = T::BITS;
+    let lhs_int = T::from(field_to_i128(lhs, width + 1)).expect("Should convert");
+    let rhs_int = T::from(field_to_i128(rhs, width + 1)).expect("Should convert");
+    let result_int = match mutation_operation {
+        BinaryIntOperationMutation::Add => lhs_int.wrapping_add(&rhs_int),
+        BinaryIntOperationMutation::Sub => lhs_int.wrapping_sub(&rhs_int),
+        BinaryIntOperationMutation::Xor => lhs_int ^ rhs_int,
+        BinaryIntOperationMutation::And => lhs_int & rhs_int,
+        BinaryIntOperationMutation::Or => lhs_int | rhs_int,
+    };
+    i128_to_field(result_int.as_(), width)
+}
+
+/// Perform a shift operation on the value and convert to field
+/// The lowest bit of direction_and_magnitude represents shift direction, higher bits - shift value
+fn shift<T: Shl<u32, Output = T> + Shr<u32, Output = T> + PrimInt + HasBits + AsPrimitive<i128>>(
+    input: &i128,
+    direction_and_magnitude: u32,
+) -> FieldElement {
+    let width = T::BITS;
+    let converted: T = T::from(*input).expect("Primitive should convert");
+    let after_update = if (direction_and_magnitude & 1u32).is_zero() {
+        converted >> (direction_and_magnitude >> 1)
+    } else {
+        converted << (direction_and_magnitude >> 1)
+    };
+    i128_to_field(after_update.as_(), width)
+}
+
+/// Convert a value from field representation to integer representation
+/// Width is needed to detect negative values in 2-complement form
+fn field_to_i128(input: &FieldElement, width: u32) -> i128 {
+    let mut initial_i128 = input.to_i128();
+    if initial_i128 >= 2i128.pow(width - 1) {
+        initial_i128 -= 2i128.pow(width);
+    }
+    initial_i128
+}
+
+/// Convert a value from integer representation to field representation
+/// Width is needed to detect negative values in 2-complement form
+fn i128_to_field(value: i128, width: u32) -> FieldElement {
+    let mut new_value = value;
+    if value < 0 {
+        new_value += 2i128.pow(width);
+    }
+    FieldElement::from(new_value)
+}
+
+/// Generate a random unsigned integer of given width and convert to field
+fn generate_random_for_width(prng: &mut XorShiftRng, width: u32) -> FieldElement {
+    FieldElement::from(prng.gen_range(0..(2i128 << width)))
```

Comment:
Is this 2 intentional here or should it be 1?

Reply:
My mistake